### PR TITLE
Added input validation for new metadata field submissions & fixed some other bugs related to metadata fields

### LINKED ISSUE #8753: Metadata schema: using a period in metadata element name should be rejected
**Describe the bug**

Right now, users can use periods as part of the metadata field element name when adding new metadata fields to the schema. 
This creates all sorts of problems and confusion all over, so the preferred route is likely to prevent periods from being used in the metadata field name.

Screenshot 2023-03-30 at 08 20 03

Screenshot 2023-03-30 at 08 20 52

**To Reproduce**
1. Add a new metadata field to the schema, where you use a period in the element name

**Expected behavior**
A validation error should be visible to the user, explaining that period is an illegal character that can't be used in a metadata field element name.

### LINKED ISSUE #8754: Metadata Schema: Editing of existing metadata fields should be limited to Scope Note only
**Describe the bug**

The following behaviour is currently in place in 7.5
- element can be edited
- qualifier can be edited
- if no qualifier is present, a new qualifier can be added on an existing field and successfully saved
- it is impossible to remove a qualifier. If you blank it out, angular will just send back the original value.

There are at least two reasons to think of why editing of existing fields in the metadata schema is problematic:
- Such a change will not be considered an item update, to the affected items that have a metadata value. However these items WILL look different in OAI (basically everywhere).
- There is no track record or provenance of what a metadata field used to be called.

Personally I think this is a gigantic can of worms that can be resolved and avoided by restricting metadata field editing capabilities on the schema's only to scope notes.

There is already deletion protection in place for metadata fields in the schema: they can only be deleted if they are not in use in any items. That, overall is very good !!!

Improved use case flows:

Replacing the use of one metadata field for another
- Create the new field
- Move all the metadata values from the old field to the new one
- Delete the old field

**To Reproduce**
1. Try to edit or remove a qualifier, or metadata field name from a field in the schema

**Expected behavior**
You shouldn't be able to do this, you should only be able to modify the scope note for a field.

**Related work**
Link to any related tickets or PRs here.

### PR DESCRIPTION
**Pull request by alexandrevryghem**
## References
* Fixes #8753
* Fixes #8754
* Related to DSpace/dspace-angular#2177
* Related to DSpace/RestContract#228

## Description
Added backend validation that checks if a dot is present in either the schema, element or qualifier. Sorted the retrieved metadata fields from the `/core/metadatafields/search/byFieldName` rest endpoint, to prevent the exact match from not being shown int the suggestion list (which lead to an error on Angular which prevents you from submitting a value for that field). And fixed pagination issues from `/core/metadatafields/search/byFieldName`.

While creating the rest contract PR I also saw that for both update schema and update metadata field the JSON parse exceptions threw `UnprocessableEntityException`, I changed this to `DSpaceBadRequestException` to be more consitent with the other endpoints.

## Instructions for Reviewers
List of changes in this PR:
* `MetadataFieldIndexFactoryImpl`
  * I added a new Solr field `fieldName_sort` that I used to sort the metadata fields retrieved by the `/core/metadatafields/search/byFieldName` rest endpoint (previously when you added more than 10 fields and completely rebuild Solr the exact match was always shown as the last field. Because Angular only checks if the field is present in the list of 10 fields who match the closest to the input you provided, and that the 11th metadata field is not send back to the user, the code marked the metadata field as invalid)
* `MetadataFieldRestRepository`
  * `findByFieldName()`
    * Replaced the value from `totalElements` with the actual value from the number of metadata fields who matched on Solr, instead of the size of the page (which is by default 10).
    * Fixed the `/core/metadatafields/search/byFieldName` pagination issues which made it impossible to change the size and the page you were on.
  * Added validation in the `createAndReturn` function to also check that the given string does not contain any dots.
  * Made `put` return 422 errors when the Element or Qualifier is being updated
* `MetadataSchemaRestRepository`
  * Added validation in the `createAndReturn` function to also check that the given string does not contain any dots.
  * Made `put` return 422 errors when the Schema name (prefix) is being updated

**Guidance for how to test and review this PR:**
- Add a lot of `dc.subject.*` fields (so that you have more than 10 fields (or 20 if you don't use the linked Angular PR)) and rebuild the whole Solr index again.
- Test the [/core/metadatafields/search/byFieldName](http://localhost:8080/server/api/core/metadatafields/search/byFieldName?schema=dc&element=subject&qualifier=&query=&exactName=&embed=schema&size=2&page=1) rest endpoint and check that you can now change the current page and the page size. Also check that the `totalElements` is correct.
- In the UI go to `Edit Item > Metadata` check that the `dc.subject` field is shown as the first field.

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Review by abollini — CHANGES_REQUESTED** [review]
hi @alexandrevryghem ,
thanks for your contribution, code looks good but the following things seem to be missing
- we need a RESTContract PR to record the agreed changes in the REST behavior (reject of special characters in creation, refuse to update anything than scope)
- I haven't found IT related to the pagination fix and the new scenario that should be checked in creation

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java [hunk 1]
```diff
@@ -142,11 +144,16 @@ protected MetadataSchemaRest put(Context context, HttpServletRequest request, St
...
-            throw new UnprocessableEntityException("Cannot parse JSON in request body", e);
+            throw new DSpaceBadRequestException("Cannot parse JSON in request body", e);
         }
 
-        if (metadataSchemaRest == null || isBlank(metadataSchemaRest.getPrefix())) {
-            throw new UnprocessableEntityException("metadata schema name cannot be blank");
+        MetadataSchema metadataSchema = metadataSchemaService.find(context, id);
+        if (metadataSchema == null) {
+            throw new ResourceNotFoundException("metadata schema with id: " + id + " not found");
+        }
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java:153]: this is not appropriate as the endpoint exists, it is a semantic error in the syntactically valid request so it should be a 422

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java [hunk 2]
```diff
@@ -93,6 +93,8 @@ protected MetadataSchemaRest createAndReturn(Context context)
         // validate fields
         if (isBlank(metadataSchemaRest.getPrefix())) {
             throw new UnprocessableEntityException("metadata schema name cannot be blank");
+        } else if (metadataSchemaRest.getPrefix().contains(".")) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java:96]: We also should check for commas, spaces and underscores... as those are all invalid in the UI.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java [hunk 3]
```diff
@@ -247,10 +260,14 @@ protected MetadataFieldRest createAndReturn(Context context)
 
         if (isBlank(metadataFieldRest.getElement())) {
             throw new UnprocessableEntityException("metadata element (in request body) cannot be blank");
+        } else if (metadataFieldRest.getElement().contains(".")) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java:263]: As I suggested in the UI PR, I think we should also check for commas and spaces.  Those should not be allowed in the element.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java [hunk 4]
```diff
@@ -247,10 +260,14 @@ protected MetadataFieldRest createAndReturn(Context context)
...
+            throw new UnprocessableEntityException("metadata element (in request body) cannot contain dots");
         }
 
         if (isBlank(metadataFieldRest.getQualifier())) {
             metadataFieldRest.setQualifier(null);
+        } else if (metadataFieldRest.getQualifier().contains(".")) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataFieldRestRepository.java:269]: As I suggested in the UI PR, I think we should also check for commas and spaces.  Those should not be allowed in the qualifier

**Review by tdonohue — CHANGES_REQUESTED** [review]
@alexandrevryghem : Overall, this looks good and it works.  I've verified it aligns with the REST Contract & tested it with the frontend PR.  In my review of the UI PR, I've suggested a few other validation checks (for spaces and commas) in element/qualifier which I've requested inline in this PR as well.

Finally, I did find what looks to be a bug in the sorting.  The sorting works *better*.  However, when there's a null qualifier (no qualifier), I'd expect those to be sorted **at the top** (before any fields which have qualifiers).  It appears they are sorted **to the bottom**, which means that they **sometimes don't appear in suggestions.**

Here's an example... searching for "dc.subject" doesn't find the existing `dc.subject` field, because it's not sorted into the first 10 results:
sorting

(But for fields where there's <10 results, it works OK.  For instance searching for "dc.type" will find `dc.type`, but it's last in the list)

Overall your sorting is an improvement.  But, it's not a complete fix yet.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @alexandrevryghem !  You were correct, I forgot to previously re-index... after reindexing, the search-by-metadata-field on the Edit Item page works properly!

I've also verified that the improved validation is all working well.  Looks great!

**Pull request merged**